### PR TITLE
CSS-10571 Implemented tests.

### PR DESCRIPTION
**Pull request by mertalpt**
## Description of Changes
This PR implements some unit and integration tests. To support the implementation of those tests and improve future maintainability, it also expands overall logging.

## Changes
- Integration test pipeline is changed.
- Documentation is updated.
- Libs are updated.
- A bug with pebble layer structure is fixed.
- `update_status` hook implementation is updated to make it more informative.
- Logs are expanded.
- Logic to cover edge cases on non-happy paths are added.
- Unit and integration tests are added.

### CODE REVIEW

**Review by kelkawi-a — APPROVED** [review]
Excellent work!

File: src/charm.py [hunk 1]
```diff
@@ -186,31 +188,60 @@ def _on_peer_relation_changed(self, event):
...
 
+        context = services.ServiceContext(self)
+        is_down = False
+        is_invalid = False
+        is_not_ready = False
         for service in SERVICES:
             if service.healthcheck is None:
                 continue
 
+            if not service.is_enabled(context):
```
[the remark this answers is not in the document]
Comment (mertalpt) [reply src/charm.py:211]: Workloads for Actions, Frontend, and GMS containers will be disabled by `pebble` according to `is_enabled` if the dependencies such as Kafka or Opensearch are not in place. So, the answer is no.

File: tests/integration/helpers.py [hunk 2]
```diff
@@ -57,3 +86,88 @@ async def deploy_charm(ops_test: OpsTest, charm: Path) -> None:
...
+    model = await ops_test.track_model(
+        alias=model_name,
+        model_name=model_name,
+        cloud_name=cloud_name,
+        use_existing=True,
+        keep=True,  # TODO (mertalpt): Switched for debug purposed, rollback before merge.
+    )
+
+    # No need for further processing.
+    if cloud_type != "k8s":
```
Comment (gtato) [inline tests/integration/helpers.py:149]: can this be `cloud_type == "lxd"`?

File: tests/integration/helpers.py [hunk 3]
```diff
@@ -57,3 +86,88 @@ async def deploy_charm(ops_test: OpsTest, charm: Path) -> None:
...
+    controller_cloud = await (await model.get_controller()).cloud()
+    if controller_cloud.cloud.type_ == "lxd":
+        code, stdout, stderr = await ops_test.juju(
+            "model-config", "-m", model.name, "workload-storage=microk8s-hostpath"
+        )
+        if code != 0:
+            logger.error("'juju model-config' failed:\n\tstdout: %s\n\tstderr: %s", stdout, stderr)
+            raise exceptions.SetupFailedError("command 'juju model-config' failed")
+
+    if enable_debug_logs:
```
Comment (gtato) [inline tests/integration/helpers.py:166]: do you need logging enabled only when cloud_type is `k8s`?

File: tests/integration/test_charm.py [hunk 4]
```diff
@@ -6,27 +6,204 @@
...
+            async with ops_test.fast_forward(fast_interval="5m"):
+                # Deploy DataHub
+                logger.info("Deploying '%s'", helpers.APP_NAME)
+                await helpers.deploy_charm(ops_test, charm)
+
+                # TODO (mertalpt): Find a way to avoid doing this.
+                # Gives `pebble_ready` time to run before making relations.
+                # I suspect it is a problem because gigabytes of OCI images are not cached
+                # and take a long time to be downloaded.
+                await asyncio.sleep(10 * 60)
```
Comment (gtato) [inline tests/integration/test_charm.py:131]: could you add this as `idle_period` in the `wait_for_idle` below?
https://pythonlibjuju.readthedocs.io/en/latest/api/juju.model.html#juju.model.Model.wait_for_idle

File: tests/integration/test_charm.py [hunk 5]
```diff
@@ -6,27 +6,204 @@
...
+
+                # Relate to offers
+                logger.info("Integrating to '%s'", helpers.KAFKA_OFFER_NAME)
+                await ops_test.model.integrate(helpers.APP_NAME, helpers.KAFKA_OFFER_NAME)
+                logger.info(
+                    "Waiting for '%s' to settle into 'blocked-idle' post '%s' integration",
+                    helpers.APP_NAME,
+                    helpers.KAFKA_OFFER_NAME,
+                )
+                await asyncio.sleep(2 * 60)
```
Comment (gtato) [inline tests/integration/test_charm.py:156]: same comment as above

**Review by gtato — APPROVED** [review]
looks good!
left a few non blocking comments

File: src/charm.py [hunk 6]
```diff
@@ -186,31 +188,60 @@ def _on_peer_relation_changed(self, event):
...
+                logger.info("service '%s' is not ready", service.name)
+                is_not_ready = True
+                continue
```
[the remark this answers is not in the document]
Comment (mertalpt) [reply src/charm.py:214]: It was like that at first, this way helps with debugging.

File: tests/integration/helpers.py (hunk 2 above)
Comment (mertalpt) [reply tests/integration/helpers.py:149]: Sure. I adapted this from Livepatch. I intend to work on the tests in the future to increase coverage and make them run faster.

File: tests/integration/helpers.py (hunk 3 above)
Comment (mertalpt) [reply tests/integration/helpers.py:166]: K8s is where DataHub is, so unless we hit something particularly nasty to debug this is what we need.

File: tests/integration/test_charm.py (hunk 4 above)
Comment (mertalpt) [reply tests/integration/test_charm.py:131]: It is not really the same thing. That option checks for continuous `idle` time. This is to allow images to get downloaded and `pebble_ready` to run. I intend to re-do the hooks (esp. for `update-status`) and the tests. This should be better afterwards.

File: tests/integration/test_charm.py (hunk 5 above)
Comment (mertalpt) [reply tests/integration/test_charm.py:156]: Similar to above, this can be replaced when status changes behave better after an overhaul.
